### ISSUE TITLE
Missing Optional in field Signature

### ISSUE THREAD

**Issue by DanLipsitt** [pydantic/pydantic#4043]
### Checks

* [x] I added a descriptive title to this issue
* [x] I have searched (google, github) for similar issues and couldn't find anything
* [x] I have read and followed [the docs](https://pydantic-docs.helpmanual.io/) and still think this is a bug

### Environment
Output of `python -c "import pydantic.utils; print(pydantic.utils.version_info())"`:
```
             pydantic version: 1.9.0
            pydantic compiled: True
                 install path: /Users/dlipsitt/Library/Application Support/pdm/venvs/parts-service---nOTk53-3.9/lib/python3.9/site-packages/pydantic
               python version: 3.9.9 (main, Dec 27 2021, 16:23:12)  [Clang 12.0.5 (clang-1205.0.22.11)]
                     platform: macOS-12.2-arm64-arm-64bit
     optional deps. installed: ['devtools', 'dotenv', 'email-validator', 'typing-extensions']
```
# Bug

Classes that inherit from BaseModel are missing `Optional` from the signature of optional fields. In comparison, Dataclasses have the behavior I would expect, which is that the introspected signature looks like the class definition:

```py
from inspect import signature
from typing import Optional

from pydantic import BaseModel
from pydantic.dataclasses import dataclass


@dataclass
class PydanticDataClassItem:
    required_val: int
    optional_val: Optional[int]
    optional_default_val: Optional[int] = 1


class PydanticItem(BaseModel):
    required_val: int
    optional_val: Optional[int]
    optional_default_val: Optional[int] = 1


def show_signature(cls):
    sig = "\n\t".join(str(val) for val in signature(cls).parameters.values())
    print(f"{cls.__qualname__}:\n\t{sig}\n")


show_signature(PydanticDataClassItem)
show_signature(PydanticItem)
```

output:
```
PydanticDataClassItem:
	required_val: int
	optional_val: Optional[int]
	optional_default_val: Optional[int] = 1

PydanticItem:
	required_val: int
	optional_val: int = None
	optional_default_val: int = 1
```

The first problem to notice here is that `optional_val: int = None` doesn't make sense because the default value in the signature is not an instance of the field's type.

# Details

The consequence that motivates this ticket is that Hypothesis isn't able to automatically infer types from Pydantic classes correctly (https://github.com/samuelcolvin/pydantic/issues/3126). 

Here's a demo where we generate filtered Hypothesis examples. The filters force the example's `optional_val` field to first be `None` and then `not None`:

```py
from hypothesis.strategies import builds


def show_filtered_examples(cls):
    filters = [
        lambda item: item.optional_val is None,
        lambda item: item.optional_val is not None,
    ]
    for fn in filters:
        print(repr(builds(cls).filter(fn).example()))


show_filtered_examples(PydanticDataClassItem)
show_filtered_examples(PydanticItem)
```

In the output, we see that we can produce examples of `PydanticDataClassItem` with either type of value. On the other hand, the attempt to produce a `not None` example of `PydanticItem` fails because it is always `None`:

```python console
PydanticDataClassItem(required_val=30949, optional_val=None, optional_default_val=1)
PydanticDataClassItem(required_val=-23293, optional_val=-23293, optional_default_val=1)
PydanticItem(required_val=-1870955627, optional_val=None, optional_default_val=1)

Traceback (most recent call last):
  File "/Users/dlipsitt/src/parts-service/test/pydantic_signature.py", line 43, in <module>
    show_filtered_examples(PydanticItem)
  File "/Users/dlipsitt/src/parts-service/test/pydantic_signature.py", line 39, in show_filtered_examples
    print(repr(builds(cls).filter(fn).example()))
  File "/Users/dlipsitt/Library/Application Support/pdm/venvs/parts-service---nOTk53-3.9/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py", line 334, in example
    example_generating_inner_function()
  File "/Users/dlipsitt/Library/Application Support/pdm/venvs/parts-service---nOTk53-3.9/lib/python3.9/site-packages/hypothesis/strategies/_internal/strategies.py", line 323, in example_generating_inner_function
    @settings(
  File "/Users/dlipsitt/Library/Application Support/pdm/venvs/parts-service---nOTk53-3.9/lib/python3.9/site-packages/hypothesis/core.py", line 1237, in wrapped_test
    raise the_error_hypothesis_found
  File "/Users/dlipsitt/Library/Application Support/pdm/venvs/parts-service---nOTk53-3.9/lib/python3.9/site-packages/hypothesis/core.py", line 817, in run_engine
    raise Unsatisfiable(f"Unable to satisfy assumptions of {rep}")
hypothesis.errors.Unsatisfiable: Unable to satisfy assumptions of example_generating_inner_function
```
(Some distracting irrelevant Hypothesis warnings have been removed.)

## Workaround
There are a couple of suggested workarounds in https://github.com/samuelcolvin/pydantic/issues/3126#issuecomment-1045229950. However, they require extra code for every optional field when using `builds`. This is unexpected and inconsistent with the way other classes (such as dataclasses, either the stdlib or Pydantic kind) work with Hypothesis.

**Comment by samuelcolvin** [comment]
PR welcome.

**Comment by ahirner** [comment]
Had a first look. Too early for a PR because related issues are unclear.

Field annotations are reconstructed from `outer_type_`:
https://github.com/samuelcolvin/pydantic/blob/cc54acb612cb5144a34caebb9ac143324a0cb4a1/pydantic/utils.py#L254-L258

`Optional[]` is erased for `field.outer_type_`: 
https://github.com/samuelcolvin/pydantic/blob/cc54acb612cb5144a34caebb9ac143324a0cb4a1/pydantic/fields.py#L628-L634

Simply preserving `outer_type_` fixes part of this issue. Yet, `test_deep_generic_with_referenced_inner_generic` fails.
https://github.com/samuelcolvin/pydantic/blob/6f46a5a1460728868f860840e3f6186aa40403d8/tests/test_generics.py#L903

I'm wondering why `Optional[]` can't be preserved, whether `type_` instead of `outer_type_` should be asserted and how #1055 might be related?

This fixes the OP's issue entirely, if the change should be entirely local:

```python
 kwargs = {'default': field.default} if not field.required and field.default is not None else {}
 # TODO: replace annotation with actual expected types once #1055 solved
 annotation = field.outer_type_ if not field.allow_none else Optional[field.outer_type_]
 merged_params[param_name] = Parameter(param_name, Parameter.KEYWORD_ONLY, annotation=annotation, **kwargs)
```

Lets clarify `test_deep_generic_with_referenced_inner_generic` first though.

**Comment by Bobronium** [comment]
I believe it might be implemented that way on purpose, since it's a call signature, not class definition signature. 

Python typing allows to have `a: int = None` in call signatures, without explicitly stating `Optional[int]`.



```shell
echo "def f(a: int = None) -> None: ..." > example.py
mypy example.py
```
```
Success: no issues found in 1 source file
```

**Comment by PrettyWood** [comment]
It's fixed in 1.10
